### PR TITLE
VC Join Request Command

### PR DESCRIPTION
**Pull request by bowenjw**
## Refactor
 - Added `dockerfile` and `docker-compose.yaml` ( to running with docker desktop)
 - Updated scripts in pakage.json to work with `dockerfile`
 - Updated module files to latest version

# VC Join Request Command
The new command `/vc request-join`, is meant to allow the automatons  movement of members to full channels with out moderator intervention. The command has one non-required options `channel`, witch may be used to target a destination channel. If no channel is given the command will try and use the channel it was run from.

Requirements to send request and approved:
- Be in a VC to send request
- Be in VC to approve Request
- Request must be approved in 10 or it will need to be sent again
- Only one user in the targeted VC needs to approve of the join request

### CODE REVIEW

**Comment by bowenjw** [comment]
the PR is missing some comments will add tomorrow also for some reason lint did not run successfully IDK why. will look in to it

File: src/Client/interactionCreate.ts [hunk 1]
```diff
@@ -44,21 +24,48 @@ export async function onInteractionCreate(interaction: Interaction) {
...
+
+				const errorMessage = 'There was an error while executing this interaction.';
+
+				// Check if interaction was deferred
+				if (interaction.deferred) {
+					// If defered interactions is followed up
+					await interaction.followUp({ content: errorMessage, ephemeral: true }).catch((e) => Logger.error(e));
+				}
+				else {
+					// Else the interactions is replied too
```
Comment (Lewdcario) [inline src/Client/interactionCreate.ts:63]: too -> to

File: locales/en-US/vc.ftl [hunk 2]
```diff
@@ -0,0 +1,21 @@
...
+request-to-join = {$member} requested to join this channel
+request-sent = Request to join {$channel} has been sent
+## Button Text
+vc-allow = Allow
+vc-reject = Reject
+## Button Responose
+appover-not-in-channel = To approve join request must be in {$channel}
+request-exspired = This request was made {$time} and has expired
+requester-not-in-old-channel = {$member} is no longer in the channel where they requested from
+move-successful = Request Approved
```
Comment (Lewdcario) [inline locales/en-US/vc.ftl:21]: Newline

File: .github/workflows/lint.yml [hunk 3]
```diff
@@ -26,21 +26,8 @@ jobs: # list of things to do
...
-              uses: actions/cache@v1
-              with:
-                  path: ${{ steps.yarn-cache-dir-path.outputs.dir }}
-                  key: ${{ runner.os }}-yarn-${{ hashFiles('**/yarn.lock') }}
-                  restore-keys: |
-                      ${{ runner.os }}-yarn-
-
             - name: Install Dependencies
-              run: yarn install --frozen-lockfile --silent
+              run: yarn install --silent
```
Comment (Lewdcario) [inline .github/workflows/lint.yml:30]: This should contain `--frozen-lockfile`

File: src/interactions/buttons/vc.ts [hunk 4]
```diff
@@ -0,0 +1,95 @@
...
+				components: []
+			})
+			.then(() => deleteMessage(interaction));
+
+		return;
+	}
+
+	// checks to see if the requester is still in the channel that they were when the request was mad
+	if (!fromChannel.members.has(args[2])) {
+		interaction
```
Comment (Lewdcario) [inline src/interactions/buttons/vc.ts:66]: Use await instead of then

File: src/interactions/buttons/vc.ts [hunk 5]
```diff
@@ -0,0 +1,95 @@
...
+					args: { member: interaction.member.toString() }
+				}),
+				components: [],
+				allowedMentions: { users: [] }
+			})
+			.then(() => deleteMessage(interaction));
+	}
+
+	// move member into requested channel
+	requester.voice.setChannel(interaction.channel as VoiceChannel);
```
Comment (Lewdcario) [inline src/interactions/buttons/vc.ts:81]: await

File: src/interactions/buttons/vc.ts [hunk 6]
```diff
@@ -0,0 +1,95 @@
...
+		.update({
+			content: t({
+				key: 'move-successful',
+				ns,
+				locale: guildLocale,
+				args: { member: interaction.member.toString() }
+			}),
+			components: []
+		})
+		.then(() => deleteMessage(interaction));
```
Comment (Lewdcario) [inline src/interactions/buttons/vc.ts:94]: await instead of then

File: src/interactions/buttons/vc.ts [hunk 7]
```diff
@@ -0,0 +1,95 @@
...
+		interaction.message.delete();
+		return;
+	}
+
+	// checks if the request was made more that 10 minutes ago
+	const date = new Date();
+	date.setMinutes(date.getMinutes() - 10);
+
+	if (interaction.createdAt < date) {
+		interaction
```
Comment (Lewdcario) [inline src/interactions/buttons/vc.ts:49]: await instead of then

File: src/interactions/buttons/vc.ts (hunk 6 above)
Comment (Lewdcario) [reply src/interactions/buttons/vc.ts:94]: On second thought, these feel redundant, so maybe combining the logic might be good

File: src/interactions/buttons/vc.ts [hunk 8]
```diff
@@ -0,0 +1,95 @@
...
+				args: { channel: interaction.channel.toString() }
+			}),
+			ephemeral: true
+		});
+		return;
+	}
+
+	// if message is rejected request is deleted
+	if (action === 'reject') {
+		interaction.message.delete();
```
Comment (Lewdcario) [inline src/interactions/buttons/vc.ts:40]: await

File: src/interactions/buttons/vc.ts [hunk 9]
```diff
@@ -0,0 +1,95 @@
...
+export default new Interaction<ButtonInteraction>().setName('vc').setExecute(async (interaction) => {
+	const { locale, guildLocale } = interaction;
+	const args = interaction.customId.split(interaction.client.splitCustomIDOn);
+	const action = args[1];
+	const requester = await interaction.guild.members.fetch(args[2]);
+	const fromChannel = (await interaction.guild.channels.fetch(args[3])) as VoiceChannel;
+
+	// checks if user responding to request is in the channel
+	if (!(interaction.channel as VoiceChannel).members.has(interaction.user.id)) {
+		interaction.reply({
```
Comment (Lewdcario) [inline src/interactions/buttons/vc.ts:26]: await

File: src/commands/chat/execution/vc/join.ts [hunk 10]
```diff
@@ -0,0 +1,123 @@
...
+				locale,
+				ns,
+				args: { channel: channel.toString() }
+			}),
+			ephemeral: true
+		});
+	}
+
+	// If channel is not full member is moved in to channel
+	interaction.member.voice.setChannel(channel);
```
Comment (Lewdcario) [inline src/commands/chat/execution/vc/join.ts:114]: await

**Review by Lewdcario — CHANGES_REQUESTED** [review]

**Comment by bowenjw** [comment]
@Lewdcario all requested changes made

File: dockerfile [hunk 11]
```diff
@@ -0,0 +1,27 @@
...
+COPY ./locales ./locales
+
+RUN yarn install --frozen-lockfile --production
+
+COPY --from=builder /usr/bot/dist/ ./dist
+COPY ./src/*.json ./dist
+
+USER node
+
+CMD [ "yarn", "start" ]
```
Comment (Lewdcario) [inline dockerfile:27]: Newline

File: src/interactions/buttons/vc.ts [hunk 12]
```diff
@@ -0,0 +1,85 @@
...
+		await interaction.message.delete();
+		return;
+	}
+
+	// checks if the request was made more that 10 minutes ago
+	const date = new Date();
+	date.setMinutes(date.getMinutes() - 10);
+
+	if (interaction.createdAt < date) {
+		updateInteraction(interaction, 'request-exspired', { time: interaction.createdAt.toDiscordString('R') });
```
Comment (Lewdcario) [inline src/interactions/buttons/vc.ts:70]: await

File: src/interactions/buttons/vc.ts [hunk 13]
```diff
@@ -0,0 +1,85 @@
...
+		return;
+	}
+
+	// checks to see if the requester is still in the channel that they were when the request was mad
+	if (!fromChannel.members.has(args[2])) {
+		updateInteraction(interaction, 'requester-not-in-old-channel', { member: interaction.member.toString() });
```
Comment (Lewdcario) [inline src/interactions/buttons/vc.ts:76]: await

File: src/interactions/buttons/vc.ts [hunk 14]
```diff
@@ -0,0 +1,85 @@
...
+		return;
+	}
+
+	// move member into requested channel
+	requester.voice.setChannel(interaction.channel as VoiceChannel);
+
+	// update message to indicat task was completed
+	updateInteraction(interaction, 'move-successful', { member: interaction.member.toString() });
```
Comment (Lewdcario) [inline src/interactions/buttons/vc.ts:84]: await

**Review by Lewdcario — CHANGES_REQUESTED** [review]
I marked some comments as unresolved as the issue persisted

**Review by Lewdcario — APPROVED** [review]

**Pull request merged**